torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+                # optimizer states follow the same strategy as
+                # the corresponding parameters
+                placemenet_strategies: List[PlacementStrategy] = []
+
+                replica_strategy = _gen_replicate_strategy(mesh)
+                placemenet_strategies.append(replica_strategy)
+
+                if node.meta["val"].ndim > 0:
+                    # if the state is a scalar, we don't want to shard it
+                    # i.e. param state step is usually a scalar tensor
+                    shard_strategy = _gen_shard_strategy(mesh, 0)
+                    placemenet_strategies.append(shard_strategy)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.STATE, placemenet_strategies
+                )
+            else:
+                activation_batch_dim_size = node.meta["val"].shape[batch_dim]
+                # find the first activation node and use its batch dim size
+                if batch_dim_analzer.batch_dim_size == -1:
+                    batch_dim_analzer.init_batch_dim_size(activation_batch_dim_size)
+
+                batch_dim_analzer.set_batch_dim(node, batch_dim)
+                shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.ACT, [shard_strategy]
+                )
+            placeholder_idx += 1
+        elif node.op == "call_function":
+            # Annotate node types for the computation graph
+            # Data Parallel node propagation logic:
+            # param (non-compute) -> out: param
+            # grad (non-compute before/after) -> out: grad
+            # state -> output: state
+            #
+            # param + activation (param must be replicate, act be sharded) -> out: activation
+            # param/state + grad (param/state/grad be the same spec) -> out: param/state
+            # param + state -> out: param
+
+            if node.target in non_compute_ops:
+                # At this point, we should have removed all the `tag_grad` nodes in the graph
+                assert node.target != torch.ops._spmd.tag_grad.default
+
+                input_nodes = node.all_input_nodes
+                assert (
+                    len(input_nodes) == 1
+                ), f"non-compute op only support one input now, found node: {node} with length of inputs: {len(node.args)}"
+                arg_node_strategy = dp_strategy_map[input_nodes[0]]
+                assert isinstance(arg_node_strategy, DataParallelStrategy)
+                arg_node_type = arg_node_strategy.node_type
+                input_full_reduction = arg_node_strategy.reduction_over_batch
+
+                if arg_node_type == NodeType.PARAM:
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.PARAM, [replica_strategy]
```

Comment:
a little confused, wouldn't Param nodes be sharded, and that implies that a non-compute op (aka a view) of the param would also be sharded?

Reply:
For `placeholders` (which are params), it could either be `replicated/sharded`, and it corresponds to DDP/FSDP style. But for `call_function` nodes (whether it's a view/non-compute or compute node), we always treat `PARAM` nodes as `Replicate`, this is because during computation it would always need the parameter be fully replicated (where it would trigger allgather operation to turn sharded param to replicated during `call_function` execution)